src/tls/Makefile.common

```diff
@@ -14,48 +37,82 @@ INCLUDE_PATHS= \
 	$(HACL_HOME)/code/lib \
 	$(KREMLIN_HOME)/kremlib \
 	$(MITLS_HOME)/libs/ffi \
-	./concrete-flags \
-        ./concrete-flags/$(FLAVOR)$(CONCRETE_FLAVOR)
+	$(SPECINC)
 
 ifeq (,$(MITLS_NODEBUG))
   INCLUDE_PATHS += ./concrete-flags/Debug
 else
   INCLUDE_PATHS += ./concrete-flags/NoDebug
 endif
 
-CACHE_DIR=cache.$(FLAVOR)$(CONCRETE_FLAVOR)
-
 FSTAR=$(FSTAR_HOME)/bin/fstar.exe --cache_dir $(CACHE_DIR) --cache_checked_modules $(OTHERFLAGS)
 
-EXTRACT_DIR=extract/$(FLAVOR)$(CONCRETE_FLAVOR)
+VFLAGS?=--use_hints --use_hint_hashes --record_hints --warn_error -271
+
+HINTS_DIR  ?= .
+CACHE_DIR   = cache/$(FLAVOR)$(CONCRETE_FLAVOR)
+EXTRACT_DIR = extract/$(FLAVOR)$(CONCRETE_FLAVOR)
+DEP_FILE    = $(CACHE_DIR)/.depend
 
-ROOTS ?= TLS.fst QUIC.fst Test.Main.fst
 
-.depend.$(FLAVOR)$(CONCRETE_FLAVOR):
-	mkdir -p $(EXTRACT_DIR)
-	$(FSTAR) --odir $(EXTRACT_DIR) \
-		 --lax \
-		 --dep full \
-		 --extract $(EXTRACT) \
-		 $(addprefix --include , $(INCLUDE_PATHS)) \
-		 $(ROOTS) > $@
+# Project Files
 
-depend: .depend.$(FLAVOR)$(CONCRETE_FLAVOR)
+# Add more roots here!
+ROOTS ?= TLS.fst QUIC.fst Test.Main.fst 
 
--include .depend.$(FLAVOR)$(CONCRETE_FLAVOR)
+EVERYTHING=$(wildcard *.fst *.fsti Make* $(MITLS_HOME)/src/pki/* $(MITLS_HOME)/libs/ffi/* ideal-flags/* concrete-flags/* concrete-flags/$(FLAVOR)$(CONCRETE_FLAVOR)/*)
 
-%.checked.lax:
-	$(FSTAR) --lax $(addprefix --include , $(INCLUDE_PATHS)) $<
-	touch $@
+
+# Production Rules
+
+$(DEP_FILE): $(EVERYTHING) 
+ifdef VERBOSE
+	@echo "\033[1;33mComputing $(FLAVOR)$(CONCRETE_FLAVOR) dependencies ...\033[;37m"
+endif
+	@mkdir -p $(EXTRACT_DIR)
+	@mkdir -p $(CACHE_DIR)
+	@mkdir -p $(HINTS_DIR)
+	$(subst \\,, \
+		$(FSTAR) --cache_dir $(CACHE_DIR) --odir $(EXTRACT_DIR) --extract $(EXTRACT) \
+			--dep full \
+			$(addprefix --include , $(INCLUDE_PATHS)) \
+			$(ROOTS) > $(DEP_FILE) \
+	)
+
+depend: $(DEP_FILE)
+
+-include $(DEP_FILE)
+
+# the touch is needed because make may decide to invoke fstar on this
+# target because the timestamp on, say, the .fst file changed, and
+# then fstar may detect that there was no change in content and not
+# re-emit the .fst.checked file. Adding the touch ensures that the
+# timestamp on the .checked file is updated to avoid make running this
+# rule again.
+
+%.checked: | $(DEP_FILE)
```

Comment:
Beware that this rule applies to files in `FStar/ulib` that `mitls-fstar` depends on. 
This means that we will re-verify these files (which seems wasteful). The `--hint_file` option will refer to a fresh `.hints` file, so we won't even use the hints recorded in `FStar`. 
Can't we just copy the `.checked` files from `FStar` (right now they aren't checked in, but perhaps they could be)?

Reply:
Yes, that's known and expected. We haven't decided about those things yet, but I think I'll simply copy all .checked files over to the caches for now.

Reply:
In yesterday's call Nik said that checking in `.checked` files in the `FStar` repo would be a bad idea, but they could be put in a separate repo like we do with `fstarlang_binaries`. 
For the time being we could just lax typecheck those files.

Reply:
Yes, we could check them into [fstar-binaries](https://github.com/FStarLang/binaries). I'm not sure they would be very useful though. We would definitely have to add some mechanism to track the right file modification times, otherwise they'd always be considered stale by either F* or make.